design-proposals/cloud-hypervisor-integration.md

```diff
@@ -0,0 +1,291 @@
+with KubeVirt.
+
+One other reason for going through this effort is to identify if the abstraction
+layers are correctly defined to support another VMM. This will help improve the
+existing code by defining cleaner interfaces if needed.
+
+## Goals
+
+Provide users a way to choose Cloud Hypervisor over libvirt/QEMU to run their
+virtual machines.
+
+## Non Goals
+
+Support all features available through KubeVirt.
+
+Since Cloud Hypervisor has a much narrower scope than libvirt/QEMU, it doesn't
+support as many features. Therefore, we can only expect a subset of KubeVirt's
+features to be supported by Cloud Hypervisor.
+
+## Definition of Users
+
+This feature is directed at KubeVirt's users who want to run virtual machine
+more securely by choosing Cloud Hypervisor over libvirt/QEMU.
+
+## User Stories
+
+A user recently tried Cloud Hypervisor and wants to use it for running virtual
+machines on his Kubernetes/KubeVirt cluster.
+
+## Repos
+
+- [KubeVirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+
+Looking at KubeVirt's architecture, each `virt-launcher` instance manages a
+single pod. This is the abstraction layer we need to replace so that instead of
+using `libvirt` to spawn QEMU's virtual machines, it will run and communicate
+with Cloud Hypervisor directly.
+
+A new launcher `ch-launcher` will be created so that it fully replaces the
+existing `virt-launcher` component when needed.
+
+## API Examples
+
+Example of how a user could request Cloud Hypervisor as the underlying
+hypervisor through the VMI spec:
+
+```yaml
+spec:
+  hypervisor: cloud-hypervisor
+```
+
+- Introduction of a new field `Hypervisor` in `VirtualMachineInstanceSpec`
+- By default if no `Hypervisor` is provided, it would default to `libvirt`.
+- The two acceptable entries would be either `cloud-hypervisor` or `libvirt`.
+
+The `virt-operator` can inform all other components about the hypervisor type
+based on the information from the VMI spec. A different `virt-launcher` image
+would be picked instead of the default one, so that it contains `ch-launcher`.
+
+## Features
+
+### Supported features
+
+Here is a list of features expected to be available with Cloud Hypervisor.
+
+#### Lifecycle
+
+- Create a VM
+- Start a VM
+- Pause/Resume a VM
+- Snapshot/Restore a VM
+- Stop a VM
+
+#### Virtual Hardware
+
+- OVMF support (EFI)
+- CPU topology
+- CPU model is exclusively the equivalent of host for QEMU (no emulation of specific CPU model)
+- RNG using virtio-rng
+- Only headless VMs as we don't have graphics or video device emulation
+- CPU constraints + hotplug
+- Memory constraints + hotplug
+- Hugepages
+
+#### NUMA
+
+- Host NUMA to select specific host CPUs and make sure memory is allocated on expected NUMA node
+- Guest NUMA to expose any NUMA configuration to the guest
+
+#### Disks and Volumes
+
+- Disk support with virtio-block or vhost-user-block
+- Volume support with virtio-fs
+

```

Comment:
You haven't mentioned here some of the storage features such has volume hotplug and online disk resize. Are the supported by CH?Otherwise, I'd listed under the unsupported features

Reply:
Ah yes good point. Volume hotplug is supported since Cloud Hypervisor can hotplug extra `virtio-block` devices. Resizing a disk isn't something that we support though. I'll update the doc accordingly.